README.md
# IMPORTANT: Latest Version

The current version is 1.0.5. Please see the [changelog](./CHANGELOG.md) for details on version history.

# What

This package implements an authentication plugin for the open-source Datastax NodeJS Driver for Apache Cassandra. The driver enables you to add authentication information to your API requests using the AWS Signature Version 4 Process (SigV4). Using the plugin, you can provide users and applications short-term credentials to access Amazon Keyspaces (for Apache Cassandra) using AWS Identity and Access Management (IAM) users and roles.

The plugin depends on the AWS SDK for NodeJS. It uses `AWSCredentialsProvider` to obtain credentials. You must specify the service endpoint to use for the connection.
You can provide the Region in the constructor programmatically, via the `AWS_REGION` environment variable.

The full documentation for the plugin is available at
[Amazon Keyspaces AWS Docs](https://docs.aws.amazon.com/keyspaces/latest/devguide/programmatic.credentials.html#programmatic.credentials.SigV4_KEYSPACES).


# Using the Plugin

The following sections describe how to use the authentication plugin for the open-source DataStax NodeJS Driver for Cassandra to access Amazon Keyspaces.

## SSL Configuration

The first step is to get an Amazon digital certificate to encrypt your connections using Transport Layer Security (TLS). The DataStax NodeJS driver must use an SSL trust store so that the client SSL engine can validate the Amazon Keyspaces certificate on connection. To use the trust store and create a certificate, see [Using a Cassandra Java Client Driver to Access Amazon Keyspaces Programmatically](https://docs.aws.amazon.com/keyspaces/latest/devguide/programmatic.drivers.html#using_java_driver).

## Region Configuration

Before you can start using the plugin, you must configure the AWS Region that the plugin will use when authenticating. This is required because SigV4 signatures are Region-specific. For example, if you are connecting to the `cassandra.us-east-2.amazonaws.com` endpoint, the Region must be `us-east-2`. For a list of available AWS Regions and endpoints, see [Service Endpoints for Amazon Keyspaces](https://docs.aws.amazon.com/keyspaces/latest/devguide/programmatic.endpoints.html).

You can specify the Region using one of the following four methods:

* Environment Variable
* System Property
* Constructor
* Configuration

## Environment Variable

You can use the `AWS_REGION` environment variable to match the endpoint that you are communicating with by setting it as part of your application start-up, as follows.

``` shell
$ export AWS_Region=us-east-1
```

## Add the Authentication Plugin to the Application

The authentication plugin supports version 4.x of the DataStax NodeJS Driver for Cassandra. To add this application use 

```bash
$ npm install aws-sigv4-auth-cassandra-plugin --save
```

## How to use the Authentication Plugin

When using the open-source DataStax NodeJS driver, the connection to your Amazon Keyspaces endpoint is represented by the `Client` class. 

### Programmatically Configure the Driver

When using the DataStax NodeJS driver, you interact with Amazon Keyspaces primarily through the `Client` class.

To use the authentication plugin, you set a Region-specific instance of SigV4AuthProvider as the authentication provider, as in the following example.

1. Create a `SigV4AuthProvider` from plugin.
1. Add an SSL context using `AmazonRootCA1.pem` ssl and Keyspaces endpoint. 
1. Set the local data center to the region name, in this example it is `us-west-2`. 
The local data center is used by the driver for routing of requests, and it is required when the builder is constructed with `addContactPoints`.
1. Set the authentication provider to a new instance of the `SigV4AuthProvider`.
You can specify the Region for the endpoints that you’re using in the constructor for `SigV4AuthProvider`, as in the following example. 
Or, you can set the environment variable or system property as shown previously.

The following code example demonstrates the previous steps.

``` js
const cassandra = require('cassandra-driver');
const fs = require('fs');
const sigV4 = require('aws-sigv4-auth-cassandra-plugin');

const auth = new sigV4.SigV4AuthProvider({
    region: 'us-west-2', 
    accessKeyId:'AKIAIOSFODNN7EXAMPLE',
    secretAccessKey: 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY'});

const sslOptions1 = {
  ca: [
      fs.readFileSync('~/.cassandra/AmazonRootCA1.pem', 'utf-8')],
  host: 'cassandra.us-west-2.amazonaws.com',
  rejectUnauthorized: true
};


const client = new cassandra.Client({
  contactPoints: ['cassandra.us-west-2.amazonaws.com'],
  localDataCenter: 'us-west-2',
  authProvider: auth,
  sslOptions: sslOptions1,
  protocolOptions: { port: 9142 }
});


const query = 'SELECT * FROM system_schema.keyspaces';

client.execute(query).then(
    result => console.log('Row from Keyspaces %s', result.rows[0]))
    .catch( e=> console.log(`${e}`));
```


CHANGELOG.md
# CHANGELOG 

## [1.0.5] - 2023-10-11
Upgrade jsdoc and mocha versions. Move jsdoc to devDependencies.

## [1.0.4] - 2022-04-08
Minor upgrade of dependent cassandra driver

## [1.0.3] - 2021-10-20 
Unify manual and default aws credential chains

## [1.0.2] - 2020-10-22
Fixed bug with temporary credentials and sessions not working


## [1.0.0] - 2020-06-18

Initial Release


test/unit/sigv4-auth-provider-tests.js
/*
 *   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

'use strict';

const AWS = require('aws-sdk');
const lib = require('../../lib/sigv4-auth-provider.js');
const assert = require('assert');

describe('SigV4AuthProvider',  () => {
  describe('#extractNonce()', () => {
    let SigV4AuthProvider = lib.SigV4AuthProvider;

    let expected = '0c0b0c6f3946d14ce1a49a8f8c86a888';

    it('should pull basic nonce=', function () {
      let buf = Buffer.from('nonce=0c0b0c6f3946d14ce1a49a8f8c86a888');
      assert.equal(SigV4AuthProvider.extractNonce(buf), expected);
    });

    it('should stop at a comma', function () {
      let buf = Buffer.from('nonce=0c0b0c6f3946d14ce1a49a8f8c86a888,,');

      assert.equal(SigV4AuthProvider.extractNonce(buf), expected);
    });

    it('should return undefined when no nonce= is present', function () {
      let buf = Buffer.from('0b0c6f3946d14ce1a49a8f8c86a888,,');

      assert.notStrictEqual(SigV4AuthProvider.extractNonce(buf), false);
    });
  });

  describe('#constructor()', () => {
    let SigV4AuthProvider = lib.SigV4AuthProvider;

    let originalFn = SigV4AuthProvider.getRegionFromEnv;
    let regionFromEnv;

    beforeEach(function () {
      regionFromEnv = "ENV_DEFAULT_REGION";
      SigV4AuthProvider.getRegionFromEnv = () => {
        return regionFromEnv
      };
    });

    afterEach(function () {
      SigV4AuthProvider.getRegionFromEnv = originalFn;
    });

    it('should use Region if Provided',  ()  => {
      let provider = new SigV4AuthProvider({region: "us-east-23", accessKeyId:'key'});

      assert.equal(provider.region, "us-east-23");
    });

    it('should use default region if Provided',  () => {
      let provider = new SigV4AuthProvider({accessKeyId:'key'});

      assert.equal(provider.region, "ENV_DEFAULT_REGION");
    });

    it('should fail if no region retrievable',  () => {
      regionFromEnv = null;

      let err = new Error(
          "[SIGV4_MISSING_REGION] No region provided.  You must either provide a region or set "
          + "environment variable [AWS_REGION]");
      assert.throws(() => {new SigV4AuthProvider()}, err);
    });

    it('should use provided credentials', (done) => {
      const credentials = new AWS.Credentials('UserID-1', 'UserSecretKey-1', 'SessiosnToken-1');
      const options = {
        region: "us-east-23",
        accessKeyId: "UserID-1",
        secretAccessKey: "UserSecretKey-1",
        sessionToken: "SessiosnToken-1"
      };
      let provider = new SigV4AuthProvider(options);
      assert.ok(provider.chain);
      assert.ok(provider.chain instanceof AWS.CredentialProviderChain);
      assert.deepEqual(provider.chain.providers, [credentials]);
      provider.chain.resolvePromise().then((creds) => {
        assert.deepEqual(creds, {
          accessKeyId: 'UserID-1',
          expireTime: null,
          expired: false,
          refreshCallbacks: [],
          sessionToken: 'SessiosnToken-1'
        });
        done();
      });
    });

    it('should use default credential chain if no access key provided', () => {
      let provider = new SigV4AuthProvider({region: "us-east-23"});
      assert.ok(provider.chain);
      assert.ok(provider.chain instanceof AWS.CredentialProviderChain);
      assert.deepEqual(provider.chain.providers, AWS.CredentialProviderChain.defaultProviders);
    });
  });
});

describe('SigV4Authenticator', () => {
  describe('#initialResponse()', () => {
    let target = new lib.SigV4AuthProvider({region: "region", accessKeyId:'key'}).newAuthenticator();

    it('should call callback function with Sigv4 buffer', () => {
      target.initialResponse((err, buf) => {
        if (err != null) {
          assert.fail("Error sent to callback");
        }

        // this is a style of buffer setup that is deprecated, however
        // it is consistent with older versions of js.  We use it
        // here as a double-entry bookkeeping that our buffer is right.
        assert.notStrictEqual(buf, new Buffer.from("SigV4\0\0", 'utf8'));
      });
    });
  });

  describe('#evaluateChallenge()', () => {
    const credentials = new AWS.Credentials('UserID-1', 'UserSecretKey-1', 'SessiosnToken-1');
    const chain = new AWS.CredentialProviderChain([credentials]);
    let target = new lib.SigV4Authenticator({
      region: 'us-west-2',
      chain: chain,
      date: new Date(1591742511000)
    });

    it('should call callback with Signed Request', (done) => {
      let nonceBuffer = Buffer.from("nonce=91703fdc2ef562e19fbdab0f58e42fe5");
      let expected = "signature=7f3691c18a81b8ce7457699effbfae5b09b4e0714ab38c1292dbdf082c9ddd87,access_key=UserID-1,amzdate=2020-06-09T22:41:51.000Z,session_token=SessiosnToken-1";

      let calledCallback = false;
      target.evaluateChallenge(nonceBuffer, (err, buff) => {
        assert.equal(buff.toString(), expected);
        done();
      });
    });

    it('should fail when Nonce is not found', () => {
      let nonceBuffer = Buffer.from("buffer1");
      let calledCallback = false;
      let expected = 'Error: [SIGV4_MISSING_NONCE] Did not find nonce in SigV4 '
          + 'challenge:[buffer1]';

      target.evaluateChallenge(nonceBuffer, (err, buff) => {
        assert.equal(expected, err.toString());
        calledCallback = true;
      });
      assert.equal(calledCallback, true);
    });
  });


});



test/unit/sigv4-auth-signature-tests.js
/*
 *   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

const CryptoJS = require ('crypto-js');
const authsign = require('../../lib/sigv4-auth-signature');
const assert = require('assert');


describe('Signing Steps',  () => {
  const signingSteps = authsign.testingOnly.signingSteps;

  // epoch milliseconds representing 2020-06-09T22:41:51.000Z.
  const requestDate = new Date(1591742511000);
  const isoDateStamp = '2020-06-09T22:41:51.000Z';
  const credentialTimeStamp = "20200609";

  const region = "us-west-2";
  const accessKeyId = 'UserID-1';
  const secretAccessKey ="UserSecretKey-1";
  const nonce = '91703fdc2ef562e19fbdab0f58e42fe5';

  const signature = '7f3691c18a81b8ce7457699effbfae5b09b4e0714ab38c1292dbdf082c9ddd87';
  const signingScope =  '20200609/us-west-2/cassandra/aws4_request';
  const canonicalRequest = 'PUT\n'
      + '/authenticate\n'
      + 'X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=UserID-1%2F20200609%2Fus-west-2%2Fcassandra%2Faws4_request&X-Amz-Date=2020-06-09T22%3A41%3A51.000Z&X-Amz-Expires=900\n'
      + 'host:cassandra\n'
      + '\n'
      + 'host\n'
      + 'ddf250111597b3f35e51e649f59e3f8b30ff5b247166d709dc1b1e60bd927070';

  const stringToSign = 'AWS4-HMAC-SHA256\n'
      + '2020-06-09T22:41:51.000Z\n'
      + '20200609/us-west-2/cassandra/aws4_request\n'
      + 'bf5cc15759befc37fc35689dcdd7938a9ad042b6326f537c4c5eecaa2ebf911e';
  const signingKey = CryptoJS.enc.Hex.parse('7fb139473f153aec1b05747b0cd5cd77a1186d22ae895a3a0128e699d72e1aba');
  const computedFullResult ='signature=7f3691c18a81b8ce7457699effbfae5b09b4e0714ab38c1292dbdf082c9ddd87,access_key=UserID-1,amzdate=2020-06-09T22:41:51.000Z';

  describe('#deriveSigningScope()', () => {
    it('should work with example above', ()  => {
      let actual = signingSteps.deriveSigningScope(credentialTimeStamp, region);

      assert.equal(actual, signingScope);
    });
  });

  describe('#deriveCannonicalRequest()', () => {
    it('should work with example above', () => {
      let nonce1 = '91703fdc2ef562e19fbdab0f58e42fe5';
      let nonceHash = CryptoJS.SHA256(nonce1);
      let actual = signingSteps.deriveCanonicalRequest(accessKeyId, signingScope,
          isoDateStamp, nonceHash);

      assert.equal(actual, canonicalRequest);
    });
  });

  describe('#createStringToSign()', () => {
    it('should format everything correctly', () => {
      let actual = signingSteps.createStringToSign(canonicalRequest, isoDateStamp,
          signingScope);

      assert.equal(actual, stringToSign);
    });
  });

  describe('#deriveSigningKey()', () => {
    it('should produce key when given valid input', () => {
      let actual = signingSteps.deriveSigningKey(secretAccessKey, credentialTimeStamp,
          region);

      assert.notStrictEqual(actual, signingKey);
    });
  });

  describe('#computeSignature()', () => {
    it('should come up with expected signature', () => {
      let actual = signingSteps.computeSignature(stringToSign, signingKey);

      assert.equal(actual.toString(), signature);
    });
  });

  describe('#formSignedString()', () => {
    it('should format everything correctly', () =>  {

      let amzDate = '2020-06-09T22:41:51.000Z';
      let actual = signingSteps.formSignedString(signature, accessKeyId, amzDate);

      assert.equal(actual, computedFullResult);
    });
  });

  describe('#computeSigV4SignatureCassandraRequest()', () => {
    it('should produce valid result when no temporary creds', () =>  {
      let actual = authsign.computeSigV4SignatureCassandraRequest({
            region: region,
            nonce: nonce,
            date: requestDate,
            accessKeyId: accessKeyId,
            secretAccessKey: secretAccessKey,
      });
      assert.equal(actual, computedFullResult);
    });
    it('should produce valid result when given temporary creds', () =>  {
      const sessionToken = 'sess-token-1';
      const computedFullResultWithSession ='signature=7f3691c18a81b8ce7457699effbfae5b09b4e0714ab38c1292dbdf082c9ddd87'
          + ',access_key=UserID-1,amzdate=2020-06-09T22:41:51.000Z'
          + ',session_token=sess-token-1';

      const actual = authsign.computeSigV4SignatureCassandraRequest({
        region: region,
        nonce: nonce,
        date: requestDate,
        accessKeyId: accessKeyId,
        secretAccessKey: secretAccessKey,
        sessionToken: sessionToken
      });
      assert.equal(actual, computedFullResultWithSession);
    });
  });
});


lib/sigv4-auth-provider.js
/*
 *   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

'use strict';

const util = require("util");
const cass = require('cassandra-driver');
const AWS = require('aws-sdk');
const sigv4 = require('./sigv4-auth-signature');

/**
 * Creates a new instance of the Authenticator provider.
 *
 * @classdesc Provides an SigV4 [Authenticator]{@link cass.auth.AuthProvider} instances to be used when
 * connecting to a host.
 * @extends cass.auth.AuthProvider
 * @param {object} options
 * @param {string} options.region - aws region such as 'us-west-2'.
 * @param {string} options.accessKeyId - if not provided default to using profile
 * @param {string} options.secretAccessKey - AWS profile iff accessKeyId is not provided.
 * @param {string} options.sessionToken - use this if you are using temporary credentials.
 * @constructor
 */
function SigV4AuthProvider( {
    region,
    accessKeyId,
    secretAccessKey,
    sessionToken} = {}) {
  if (accessKeyId) {
    const credentials = new AWS.Credentials(accessKeyId, secretAccessKey, sessionToken);
    this.chain = new AWS.CredentialProviderChain([credentials]);
  } else {
    this.chain = new AWS.CredentialProviderChain();
  }

  this.region = region || SigV4AuthProvider.getRegionFromEnv();

  // validate region
  if (!this.region) {
    throw new Error('[SIGV4_MISSING_REGION] No region provided.  You must either provide a region or set '
    + 'environment variable [AWS_REGION]');
  }
}

util.inherits(SigV4AuthProvider, cass.auth.AuthProvider);

/**
 * Pull region from enviornment
 * @returns {string}
 */
SigV4AuthProvider.getRegionFromEnv = () => {
  return process.env.AWS_REGION;
};

/**
 * Pull the nonce from a challenge buffer
 *
 * @param {Buffer} buf - Buffer with service response.
 * @returns {string} - should be the string representing the nonce=.
 */
SigV4AuthProvider.extractNonce = (buf) => {
  let bufAsString = buf.toString();

  let res1 = bufAsString.split("nonce=");

  if (res1.length < 2) {
    return undefined;
  }

  let res2 = res1[1].split(',');

  return res2[0];
};


/**
 * Returns a new [Authenticator] instance to
 * be used for SigV4 authentication.
 * @override
 * @returns {Authenticator}
 */
SigV4AuthProvider.prototype.newAuthenticator = function () {
  return new SigV4Authenticator({
    region: this.region,
    chain: this.chain,
  });
};

/**
 * Creates a new instance of the Authenticator for SigV4.
 *
 * Generally speaking you should avoid constructing this directly, and instead
 * really on {@link SigV4AuthProvider} newAuthenticator method
 *
 * @classdesc allows SigV4 to be used as an authentication method.
 * @extends cass.auth.Authenticator
 * @param {object} options
 * @param {string} options.region - aws region such as 'us-west-2'.
 * @param {AWS.CredentialProviderChain} options.chain - provider chain with appropriate credentials
 * @param {Date} options.date - fixed date to use.  If not provided, we use current date.
 * @constructor
 */
function SigV4Authenticator({
  region,
  chain,
  date
} = {}) {
  this.chain = chain;
  this.region = region;
  this.date = date;
}

util.inherits(SigV4Authenticator, cass.auth.Authenticator);

/**
 * Initially called when validation is performed.
 *
 * @param {function} callback - called when sending desired authentication
 * indicator
 */
SigV4Authenticator.prototype.initialResponse = function (callback) {
  // we need to tell the system we want sigV4.
  const responseBuffer = Buffer.from("SigV4\0\0", 'utf8');
  callback(null, responseBuffer);

};

/**
 * Called when Service responds with a challenge request.
 * We are then responsible for extracting the nonce, and providing a SigV4
 * signature.
 * @param {Buffer} challenge - buffer of challenge
 * @param {function} callback - function used with existing credentials.
 */
SigV4Authenticator.prototype.evaluateChallenge = function (challenge, callback) {
  let nonce = SigV4AuthProvider.extractNonce(challenge);
  if (!nonce) {
    callback(new Error(`[SIGV4_MISSING_NONCE] Did not find nonce in SigV4 challenge:[${challenge}]`), null);
    return;
  }

  let dateToUse  = this.date || new Date();

  this.chain.resolvePromise().then((creds) => {
    let signedString =  sigv4.computeSigV4SignatureCassandraRequest({
      region: this.region,
      accessKeyId: creds.accessKeyId,
      secretAccessKey: creds.secretAccessKey,
      sessionToken: creds.sessionToken,
      date: dateToUse,
      nonce: nonce
    });
    callback(null, Buffer.from(signedString));
  });
};

module.exports =
{
  SigV4AuthProvider: SigV4AuthProvider,
  SigV4Authenticator: SigV4Authenticator
};


lib/sigv4-auth-signature.js
/*
 *   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

'use strict';

let CryptoJS = require ('crypto-js');

const CASSANDRA_SERVICE_NAME = 'cassandra';
const AWS4_SIGNING_ALGORITHM = 'AWS4-HMAC-SHA256';
const V4_IDENTIFIER = 'aws4_request';

/**
 * Compute the signature from signing key.
 *
 * @param {string} stringToSign - the complete string we must sign
 * @param {WordArray} signingKey - derived from date, region etc
 * @returns {WordArray} constituting the signture
 * @private
 */
function computeSignature(stringToSign, signingKey) {
  return  CryptoJS.HmacSHA256(stringToSign, signingKey);
}

/**
 * Form the authentication string which contains metadata and signature.
 *
 * @param {WordArray | string} signature - cypher-text for the request
 * @param {string} accessKeyId - Access id of AccessKey pair
 * @param {string} isoDateString - timestamp of form '2020-06-09T22:41:51.000Z'
 * @param {string} sessionToken - session token if temporary credentials
 * @returns {string} full authentication string
 * @private
 */
function formSignedString(signature, accessKeyId, isoDateString, sessionToken) {
  let result = `signature=${signature},access_key=${accessKeyId},amzdate=${isoDateString}`;

  if (sessionToken) {
    result += `,session_token=${sessionToken}`;
  }

  return result;
}

/**
 * Creates the signing key for signature.
 * This is based on the date and region
 *
 * @param {string} secretAccessKey - key from AccessKey pair
 * @param {string} credentialDateStamp - aws credential stamp, '20200609' for example
 * @param {string} region - the aws region, example 'us-west-2'
 * @returns {WordArray} binary chunk representing the signing key.
 * @private
 */
function deriveSigningKey(secretAccessKey, credentialDateStamp, region) {
  let secret = 'AWS4' + secretAccessKey;
  let dateHmac =  CryptoJS.HmacSHA256(credentialDateStamp, secret);
  let regionHmac = CryptoJS.HmacSHA256(region, dateHmac);
  let serviceHmac = CryptoJS.HmacSHA256(CASSANDRA_SERVICE_NAME, regionHmac);

  return  CryptoJS.HmacSHA256("aws4_request", serviceHmac);
}

/**
 * Transforms a date into an aws credential date stamp.
 *
 * @example 2020-06-09T22:41:51.000Z -> '20200609'
 * @param {Date} date - representing the request time
 * @returns {string} aws credential timestamp
 * @private
 */
function toCredentialDateStamp(date) {
  let result = date.toISOString().replace(/[:\-]|\.\d{3}/g, '');

  return result.substring(0, 8);
}

/**
 * Form the data that will be checked against the signature we build.
 *
 * @param {string} canonicalRequest - the formal request sorted, and made
 * unambiguous
 * @param {string} isoDateString - timestamp of form '2020-06-09T22:41:51.000Z'
 * @param {string} signingScope - description defining the request
 * @returns {string} the string that will be compared against to ensure
 * authentication
 * @private
 */
function createStringToSign(canonicalRequest, isoDateString, signingScope) {
  let digest = CryptoJS.SHA256(canonicalRequest);

  return `${AWS4_SIGNING_ALGORITHM}\n${isoDateString}\n${signingScope}\n${digest}`;
}

/**
 * Determines a scope string that is used in the plain text to be signed.
 *
 * @param {string} credentialDateStamp - aws credential stamp, '20200609' for example
 * @param {string} region - aws region such as 'us-west-2'
 * @returns {string} for example '20200609/us-west-2/cassandra/aws4_request'
 * @private
 */
function deriveSigningScope(credentialDateStamp, region) {
  return [
    credentialDateStamp,
    region,
    CASSANDRA_SERVICE_NAME,
    V4_IDENTIFIER
  ].join('/');
}

function formatXAmzCred(accessKeyId, scope) {
  return `X-Amz-Credential=${accessKeyId}%2F${encodeURIComponent(scope)}`;
}

function formatXAmzDate(timestamp) {
  return `X-Amz-Date=${encodeURIComponent(timestamp)}`
}

const ADZ_ALGORITHM_HEADER = `X-Amz-Algorithm=${AWS4_SIGNING_ALGORITHM}`;
const AMZ_EXPIRES_HEADER = "X-Amz-Expires=900";

/**
 * Creates the canonical request.  This is a sorted, unambiguous version of
 * the request that will be compared to for authentication.
 *
 * @param {string} accessKeyId - access id of the AccessKey pair
 * @param {string} signingScope - description defining the request
 * @param {string} isoDateString - timestamp of form '2020-06-09T22:41:51.000Z'
 * @param nonceHash - sha256 digest of the nonce provided in the challenge buffer
 * @returns {string} the canonical request.
 * @private
 */
function deriveCanonicalRequest(accessKeyId, signingScope, isoDateString, nonceHash) {
  let headers = [
    ADZ_ALGORITHM_HEADER,
    formatXAmzCred(accessKeyId, signingScope),
    formatXAmzDate(isoDateString),
    AMZ_EXPIRES_HEADER];

  headers.sort();

  let queryString = headers.join('&');

  return `PUT\n/authenticate\n${queryString}\nhost:${CASSANDRA_SERVICE_NAME}\n\nhost\n${nonceHash}`;

}

/**
 * Computes the signature line of a given cassandra request.
 *
 * @param {object} options
 * @param {string} options.region - region such as 'us-west-2'
 * @param {string} options.nonce - nonce provided by the challenge request
 * @param {Date} options.date - date representing the time of the request
 * @param {string} options.accessKeyId - access id of the AccessKey pair
 * @param {string} options.secretAccessKey - password/secret of the AccessKey pair
 * @param {string} options.sessionToken - optionally set when access credentials are
 * temporary.
 * @returns {string} a complete signature string.
 * @example
 * // returns
 * // 'signature=7f3691c18a81b8ce7457699effbfae5b09b4e0714ab38c1292dbdf082c9ddd87,access_key=UserID-1,amzdate=2020-06-09T22:41:51.000Z'
 * let response = computeSigV4SignatureCassandraRequest({
           region: 'us-west-2',
           nonce: '91703fdc2ef562e19fbdab0f58e42fe5',
           date: new Date(1591742511000),
           accessKeyId: 'UserID-1',
           secretAccessKey: 'UserSecretKey-1'
       });
  */
function computeSigV4SignatureCassandraRequest({
  region,
  nonce,
  date,
  accessKeyId,
  secretAccessKey,
  sessionToken
}) {

  let isoDate = date.toISOString();
  let credentialDateStamp = toCredentialDateStamp(date);
  let nonceHash = CryptoJS.SHA256(nonce);

  let signingScope = deriveSigningScope(credentialDateStamp, region);
  let cannoicalRequest = deriveCanonicalRequest(accessKeyId, signingScope, isoDate, nonceHash);

  let signingKey = deriveSigningKey(secretAccessKey, credentialDateStamp, region);
  let stringToSign = createStringToSign(cannoicalRequest, isoDate, signingScope);
  let signature = computeSignature(stringToSign, signingKey);

  return formSignedString(signature, accessKeyId, isoDate, sessionToken);
}

/**
 *  This is a namespace style object solely used to define the steps of the signing process
 *  While normally its not the best id to expose internals or to test them, because
 *  this function makes successive digests, sha hashes and encryption its easy to
 *  be off by a single property, type etc and not know where the problem occurred.
 *  this allows us to confirm each step of the signing process with a test, so
 *  we can quickly find a problem.
 *  @private
 */
const testingOnly = {
  signingSteps : {
    deriveSigningScope: deriveSigningScope,
    deriveCanonicalRequest: deriveCanonicalRequest,
    deriveSigningKey: deriveSigningKey,
    createStringToSign: createStringToSign,
    computeSignature: computeSignature,
    formSignedString: formSignedString
  }
};

module.exports = {
  computeSigV4SignatureCassandraRequest: computeSigV4SignatureCassandraRequest,
  testingOnly: testingOnly
};
